web/apps/serve.go

```diff
@@ -342,6 +337,56 @@ func buildServeParams(
 	}
 }
 
+func getServeToken(
+	c echo.Context,
+	inst *instance.Instance,
+	webapp *app.WebappManifest,
+	isLoggedIn bool,
+	sessID string,
+) string {
+	sharecode := c.QueryParam("sharecode")
+	if sharecode == "" {
+		if isLoggedIn {
+			return inst.BuildAppToken(webapp.Slug(), sessID)
+		}
+		return ""
+	}
+
+	// XXX The sharecode can be used for share by links, or for Cozy to Cozy
+	// sharings. When it is used for a Cozy to Cozy sharing, it can be for a
+	// preview token, and we need to upgrade it to an interact token if the
+	// member has a known Cozy URL. We do this upgrade when serving the preview
+	// page, not when sending the invitation link by mail, because we want the
+	// same link to work (and be upgraded) after the user has accepted the
+	// sharing.
+	token, pdoc, err := permission.GetTokenAndPermissionsFromShortcode(inst, sharecode)
+	if err != nil || pdoc.Type != permission.TypeSharePreview {
+		return sharecode
+	}
+	sharingID := strings.Split(pdoc.SourceID, "/")
```

Comment:
nit: Since the actual sharing ID is stored in the second part of the split, I think we should assign it directly to `sharingID`.